Third-party-libraries.md

```diff
@@ -143,3 +158,98 @@ We download and install the Redis CLI and Elasticsearch development server. We p
 ### `install_backend_python_libs.py`
 
 This script installs the Python dependencies we need in production to `third_party/python_libs`. We define these dependencies using `requirements.in`, which lists the libraries we depend on directly. Then `install_backend_python_libs.py` runs `scripts.regenerate_requirements` to generate `requirements.txt`, which lists those direct dependencies, plus all the packages that our direct dependencies need, and so on. Both `requirements.in` and `requirements.txt` specify versions, so `requirements.txt` is analogous to `yarn.lock` in that it pins all the versions of all the Python packages we use in production.
+
+## Upgrade dependencies
+
+### Upgrade backend libraries
+
+Before upgrading any backend dependencies, you should check for any breaking changes between the currently installed version and the new version you want to upgrade the library to. You should:
+
+* Check the library's changelog for breaking changes.
+* Test that after you install the upgraded version (see below), everything works correctly.
+
+Note that we don't have an automatic way to detect outdated backend dependencies, so to upgrade all outdated backend dependencies, you have to check each version manually.
+
+#### Upgrade production, backend libraries installed using pip
+
+A production library installed using pip can be upgraded as follows:
+
+1. Update the library's version number in `requirements.in`.
+2. Run `scripts/regenerate_requirements.py` to update `requirements.txt` based on the new `requirements.in`.
+
+#### Upgrade development, backend libraries
+
+A development library installed using pip can be upgraded by changing the version specified in our `scripts/install_third_party_libs.py` script. Versions are specified in one of the following places:
+
+* The `PREREQUISITES` constant.
+* The `local_pip_dependencies` variable.
+* The `system_pip_dependencies` variable.
+
+Note that for some of these locations, a constant from `scripts/common.py` may be used to specify the version.
+
+Backend libraries that are not installed using pip also have their versions specified in `install_third_party_libs.py`, but they use custom code for each library. You should read the code to find the section for your library.
+
+### Upgrade frontend libraries
+
+#### Upgrade frontend libraries installed from npm
+
+You can update all frontend libraries that we install from npm as follows. Note that `<yarn version>` specifies the currently-installed version of yarn (there should only be one version in `oppia_tools`). Also, all commands should be run from the repository root.
+
+1. Run `../oppia_tools/yarn-<yarn version>/bin/yarn upgrade`
```

Comment:
This upgrades libraries that should not have breaking changes (it works quite reliably) the second step then shows only libraries that might contain breaking changes. Probably worth mentioning here.

Reply:
done